Airthings BLE unique id migration

LaStrada opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Need to migrate `unique_id` from `<name>_<type_of_sensor>` to `<ble_address>_<type_of_sensor>`. This will make everything easier, including maintaining this integration. Unfortunately I introduced a bug in [this PR](https://github.com/home-assistant/core/pull/97905), which caused new entities to be changed.

This PR will try to migrate the entities from pre 2023.9 release. If not found the entities introduced in 2023.9.0 will be migrated. And if it finds the _newest_ and correct unique id, it will just skip the migration.

Also adding identifiers to the device, which was completely missing.

Should probably have made some tests for this, but I'm not 100% sure how to test the migration.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #99786
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

LaStrada committed:
Fix sensor unique id

LaStrada committed:
Add sensor identifiers

LaStrada committed:
Migrate entities to new unique id

LaStrada committed:
Fix linting issues

LaStrada committed:
Fix crash when migrating entity fails

LaStrada committed:
Merge branch 'dev' into airthings-ble-unique-id-migration

bdraco commented:
I'd try to migrate the older ones first since they will have the most history

LaStrada commented:
> I'd try to migrate the older ones first since they will have the most history

From where can I get that information?

bdraco commented:
Try fetching the entity registry with the 2023.8.x format first, if that fails try the 2023.9.0 format

https://github.com/home-assistant/core/blob/2a3ebbc26cb4f09d561d5bde71ef420297b6744e/homeassistant/helpers/entity_registry.py#L538

mkmer reviewed (approved):
LGTM

LaStrada committed:
Change how entities are migrated

LaStrada committed:
Merge branch 'dev' into airthings-ble-unique-id-migration

LaStrada committed:
Remve debug logging

LaStrada commented:
PR updated, and updated the PR description as well to reflect the changes.

LaStrada committed:
Remove unneeded async

LaStrada committed:
Remove migration code from init file

LaStrada committed:
Add migration code to sensor.py

LaStrada committed:
Adjust for loops to improve speed

LaStrada committed:
Bugfixes, improve documentation

LaStrada committed:
Remove old comment

LaStrada committed:
Merge branch 'dev' into airthings-ble-unique-id-migration

LaStrada committed:
Remove unused function parameter

LaStrada commented:
Refactored the code a bit. Moved the `unique_id` migration from `__init__.py` to `sensor.py` so I could drop the regex to find the sensor type (end of the `unique_id`). This also makes the loop much easier to read, less "magic" (regex) and it migrates one sensor type at the time (e.g. temperature) instead of going through every sensor.

bdraco commented on homeassistant/components/airthings_ble/sensor.py:
```diff
@@ -107,9 +115,65 @@
 }
 
 
+def migrate(hass: HomeAssistant, address: str, sensor_name: str):
```
```suggestion
@callback
def async_migrate(hass: HomeAssistant, address: str, sensor_name: str):
```

Add `async_` to callback function to indicate they are safe to run in the event loop and do not do blocking I/O

bdraco commented on homeassistant/components/airthings_ble/sensor.py:
```diff
@@ -107,9 +115,65 @@
+
+    unique_ids: dict[str, str] = {}
+
+    for entity in filtered_entities:
+        if entity.unique_id.startswith(address):
+            unique_ids["v3"] = entity.entity_id
+        elif "(" in entity.unique_id:
+            unique_ids["v2"] = entity.entity_id
+        else:
+            unique_ids["v1"] = entity.entity_id
+
+    def _migrate_unique_id(entity_id: str, new_unique_id: str):
+        ent_reg.async_update_entity(entity_id=entity_id, new_unique_id=new_unique_id)
+        _LOGGER.info("Migrated entity '%s' to unique id '%s'", entity_id, new_unique_id)
+
+    if unique_ids.get("v3"):
+        # Already has the newest unique id format
+        return
+
+    new_unique_id = f"{address}_{sensor_name}"
+
+    # Try to migrate the pre 2023.9.0 unique id first
+    if entity_id := unique_ids.get("v1"):
+        _migrate_unique_id(
+            entity_id=entity_id,
+            new_unique_id=new_unique_id,
+        )
+
```
Since there is so much branching here, can you add a test for migrating the ID from both formats?

LaStrada committed:
Address PR feedback

LaStrada committed:
Add tests

bdraco commented on tests/components/airthings_ble/test_sensor.py:
```diff
@@ -0,0 +1,169 @@
+        device_id=device.id,
+    )
+
+    # Migrate the entities
+    async_migrate(hass, WAVE_SERVICE_INFO.address, "temperature")
+
+    entities = async_entries_for_device(
+        entity_registry,
+        device_id=device.id,
+    )
```
This is great 👍 , but we need it to be an end-to-end test and which does not interact with any integration details

https://developers.home-assistant.io/docs/development_testing?_highlight=wri#writing-tests-for-integrations
> Make sure to not interact with any integration details in tests of integrations. Following this pattern will make the tests more robust for changes in the integration.

You can accomplish this by setting up the integration once the mocked entities are in the registry, after its finished setup, check the registry again to make sure the migration did what you expected.

LaStrada replied:
I believe there is room for further improvement in the tests, but they currently undergo end-to-end testing (as far as I know).

LaStrada committed:
Improve tests and test data

LaStrada committed:
Merge branch 'dev' into airthings-ble-unique-id-migration

LaStrada commented on tests/components/airthings_ble/test_sensor.py:
```diff
@@ -0,0 +1,256 @@
+
+    await hass.async_block_till_done()
+
+    with patch(
+        "homeassistant.components.airthings_ble.DataUpdateCoordinator",
+        return_value=DataUpdateCoordinator(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_method=_async_update_method,
+            update_interval=timedelta(seconds=0),
+        ),
+    ):
```
I'm was getting a `Unable to fetch data: path` from this line when I didn't patch anything:
https://github.com/home-assistant/core/blob/dev/homeassistant/components/airthings_ble/__init__.py#L48

Not sure why, or how to fix. Tried to inject BLE services so everything can be generated as it was a running system, but I wasn't able to get this working.

bdraco replied:
I think you should be able to patch `homeassistant.compnents.airthing_ble.AirthingsBluetoothDeviceData.update_device` to return a mocked `AirthingsDevice`

LaStrada replied:
Yes, it was as simple as that... Really grateful for the feedback! Will update the PR!

LaStrada committed:
Refactor test

bdraco commented on homeassistant/components/airthings_ble/sensor.py:
```diff
@@ -107,9 +115,66 @@
+
+    def _migrate_unique_id(entity_id: str, new_unique_id: str):
+        ent_reg.async_update_entity(entity_id=entity_id, new_unique_id=new_unique_id)
+        _LOGGER.info("Migrated entity '%s' to unique id '%s'", entity_id, new_unique_id)
```
```suggestion
        _LOGGER.debug("Migrated entity '%s' to unique id '%s'", entity_id, new_unique_id)
```

LaStrada committed:
Update logger level

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco commented on homeassistant/components/airthings_ble/sensor.py:
```diff
@@ -107,9 +115,66 @@
+    device_registry = device_async_get(hass)
+    entity_registry = entity_async_get(hass)
+
+    device = device_registry.async_get_device(identifiers={(DOMAIN, address)})
+
+    if not device:
+        return
```
We can lookup the device by connection and verify its the right config entry instead of adding the identifiers

LaStrada replied:
Was this what you asked for..? Or did I miss something? I changed from `identifiers` to `connections`.

bdraco replied:
All good now 👍

bdraco commented on homeassistant/components/airthings_ble/sensor.py:
```diff
@@ -107,9 +115,66 @@
+    device_registry = device_async_get(hass)
+    entity_registry = entity_async_get(hass)
+
+    device = device_registry.async_get_device(identifiers={(DOMAIN, address)})
```
You should be able to fetch by connections

bdraco commented on homeassistant/components/airthings_ble/sensor.py:
```diff
@@ -165,12 +231,18 @@ def __init__(
                     airthings_device.address,
+                ),
+            },
+            identifiers={
+                (
+                    DOMAIN,
+                    airthings_device.address,
                 )
             },
```
```suggestion
```

We generally don't add identifiers when we already have connections

LaStrada committed:
Adjust PR comments

bdraco commented on homeassistant/components/airthings_ble/sensor.py:
```diff
@@ -107,9 +115,68 @@
+    device_registry = device_async_get(hass)
+    entity_registry = entity_async_get(hass)
+
+    device = device_registry.async_get_device(connections={(DOMAIN, address)})
```
```suggestion
    device = device_registry.async_get_device(connections={(CONNECTION_BLUETOOTH, address)})
```

bdraco commented on homeassistant/components/airthings_ble/sensor.py:
```diff
@@ -165,13 +233,13 @@ def __init__(
                     CONNECTION_BLUETOOTH,
                     airthings_device.address,
-                )
+                ),
```
```suggestion
                )
```

LaStrada committed:
Address more PR comments

bdraco commented on homeassistant/components/airthings_ble/sensor.py:
```diff
@@ -107,9 +115,70 @@
+        include_disabled_entities=True,
+    )
+
+    filtered_entities = (
+        entity for entity in entities if entity.unique_id.endswith(sensor_name)
+    )
+
+    unique_ids: dict[str, str] = {}
+
+    for entity in filtered_entities:
+        if entity.unique_id.startswith(address):
+            unique_ids["v3"] = entity.entity_id
+        elif "(" in entity.unique_id:
+            unique_ids["v2"] = entity.entity_id
+        else:
+            unique_ids["v1"] = entity.entity_id
+
+    def _migrate_unique_id(entity_id: str, new_unique_id: str):
+        ent_reg.async_update_entity(entity_id=entity_id, new_unique_id=new_unique_id)
+        _LOGGER.debug(
+            "Migrated entity '%s' to unique id '%s'", entity_id, new_unique_id
+        )
+
+    if unique_ids.get("v3"):
+        # Already has the newest unique id format
+        return
+
+    new_unique_id = f"{address}_{sensor_name}"
+
+    # Try to migrate the pre 2023.9.0 unique id first
+    if entity_id := unique_ids.get("v1"):
+        _migrate_unique_id(
+            entity_id=entity_id,
+            new_unique_id=new_unique_id,
+        )
+
+    # If pre 2023.9.0 unique id is not found, try to migrate the unique id
+    # introduced in 2023.9.0
+    elif entity_id := unique_ids.get("v2"):
+        _migrate_unique_id(entity_id=entity_id, new_unique_id=new_unique_id)
```
```suggestion
@callback
def async_migrate(hass: HomeAssistant, address: str, sensor_name: str) -> None:
    """Migrate entities to new unique ids (with BLE Address)."""
    ent_reg = entity_async_get(hass)
    unique_id_trailer = f"_{sensor_name}"
    new_unique_id = f"{address}{unique_id_trailer}"
    if ent_reg.async_get_entity_id(DOMAIN, Platform.SENSOR, new_unique_id):
        # New unique id already exists
        return

    dev_reg = device_async_get(hass)
    if not (
        device := dev_reg.async_get_device(
            connections={(CONNECTION_BLUETOOTH, address)}
        )
    ):
        return

    entities = async_entries_for_device(
        ent_reg,
        device_id=device.id,
        include_disabled_entities=True,
    )
    matching_reg_entry: RegistryEntry | None = None
    for entry in entities:
        if entry.unique_id.endswith(unique_id_trailer):
            matching_reg_entry = entry
            break

    if not matching_reg_entry or matching_reg_entry.unique_id.startswith(address):
        # Already has the newest unique id format
        return

    entity_id = matching_reg_entry.entity_id
    ent_reg.async_update_entity(entity_id=entity_id, new_unique_id=new_unique_id)
    _LOGGER.debug("Migrated entity '%s' to unique id '%s'", entity_id, new_unique_id)
```

I think this can be simplified a bit

LaStrada replied:
Nice! But are we guaranteed that the oldest entry will be found first and used? If I change the order in the test to add the newest format first, then the old format ([these lines](https://github.com/home-assistant/core/pull/99832/files?diff=unified&w=0#diff-e05d3e97d19748dccce281e7885ea7393e73d7471e45e09d829a8b19ae8b2a3fR120-R134)), the test will fail. The difference between my code and yours is that my code will try to find the oldest (v1), then the 2023.9.0 format (v2). Your code doesn't seem to take care of that. If we are guaranteed that the oldest (v1) format will be found first and used, this will work, and we can skip some logic

bdraco replied:
It won't currently but it can be adjusted to not `break` in and let it keep going so the oldest one wins there.
```python
    for entry in entities:
        if entry.unique_id.endswith(unique_id_trailer):
            matching_reg_entry = entry
            break
```

I guess we need a test for this case

LaStrada replied:
Updated the code and changed the logic a bit:

If `entry.unique_id` does _not_ contain a `(` it means it's the format from before 2023.9.0, and we will try to migrate that one.

If `matching_reg_entry` is None we will still save whatever we find of `unique_id`. The result will be the v1 format if it exists.

Also removed the check to see if it finds a entry with the newest format, because that was already done earlier:
`if ent_reg.async_get_entity_id(DOMAIN, Platform.SENSOR, new_unique_id):`

LaStrada committed:
Address PR comments and adjust tests

bdraco commented on homeassistant/components/airthings_ble/sensor.py:
```diff
@@ -107,9 +117,42 @@
+    matching_reg_entry: RegistryEntry | None = None
+    for entry in entities:
+        if entry.unique_id.endswith(unique_id_trailer):
+            if not matching_reg_entry or "(" not in entry.unique_id:
```
This if can be combined with the if above since there is no branching

LaStrada committed:
Fix PR comment

bdraco reviewed (approved):
Thanks @LaStrada

MartinHjelmare commented on tests/components/airthings_ble/test_sensor.py:
```diff
@@ -0,0 +1,213 @@
+
+    assert (
+        entity_registry.async_get(sensor.entity_id).unique_id
+        == WAVE_DEVICE_INFO.address + "_temperature"
```
Please use f-strings instead of string concatenation.

MartinHjelmare commented on tests/components/airthings_ble/test_sensor.py:
```diff
@@ -0,0 +1,213 @@
+        entity_registry.async_get(v1.entity_id).unique_id
+        == WAVE_DEVICE_INFO.address + "_co2"
+    )
+    assert entity_registry.async_get(v2.entity_id).unique_id == v2.unique_id
```
I'd store the expected unique id in a local variable so we don't rely on the details of the entity registry implementation creating new entity registry entry instances on update. If that would not happen, this check would be useless as it would compare the same attribute on the same instance.
